### PR TITLE
don't return logits for benchmark script

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151075

PT2 benchmark scripts has a pattern like:
```
    def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
        cloned_inputs = clone_inputs(inputs)
        self.optimizer_zero_grad(mod)
        with self.autocast(**self.autocast_arg):
            pred = mod(**cloned_inputs)
            loss = self.compute_loss(pred)
        self.grad_scaler.scale(loss).backward()
        self.optimizer_step()
        if collect_outputs:
            return collect_results(mod, pred, loss, cloned_inputs)
        return None
```
for training.

The collect_outputs argument is True only for accuracy testing and it's false for performance testing.

For HF benchmark suite, a model usually returns tuple (loss, logits). For performance testing, even though the logits is never used anywhere, dynamo has to keep it due to the control flow.

A few bad things if we keep logits here
1. the peak memory will be higher since the logits is large and we can not release its memory earlier.
2. we can not do optimization like chunking for the logits because the tensor needs to be returned from the pre-grad graph

Actually I think it's fine to not return logits at all. 
- For training cases, checking loss and gradients for accuracy is good enough. It's hard to see two runs have mismatch logits but matching loss/gradients.
- Also, discarding logits as soon as possible for perf benchmarking makes it more fair for us.

On the other hand, it may be interesting to let dynamo support something like dynamo.constexpr (similar to tl.constexpr). A variable annotated as dynamo.constexpr will be specialized at compile time and we can do more optimization (DCE e.g.) at compile time. (A small [repro](https://gist.github.com/shunting314/0912a8947028a904c34f361021b8024d))

Benchmark results here [link](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Fri%2C%2004%20Apr%202025%2018%3A03%3A26%20GMT&stopTime=Fri%2C%2011%20Apr%202025%2018%3A03%3A26%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=gh/shunting314/204/head&lCommit=fe25dab3f65e1b0e9db0af03f7664af70fcc9c66&rBranch=main&rCommit=55e62ff74ad5614faf80b060c7bfc551e3b7af5a)
- HF 15% (1.51 -> 1.66 compression ratio) peak memory improvement
- I also see 5% (2.74 -> 2.79x) perf win for HF. It could be true. We may generate more efficient kernels since we don't need keep logits and return it from the pre-grad graph. But I'll double check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by eellison — APPROVED** [review]
Nice ! I'm not sure that constexpr  is the right term here.. but maybe unused() ? Or they could rewrite user code..

**Review by jansel — APPROVED** [review]

**Comment by shunting314** [comment]
Confirmed the metrics with local runs.

E.g. for DistillGPT2, the model will return logits as well as key/value tensors for each layer. Before the fix, the pre-grad graph needs to return the all the tensors mentioned above.
- latency with torch compile 23.78ms or 23.97ms (2 runs), peak memory 9.14GB

With the fix, the pre-grad graph only needs to return loss
- latency with torch compile 22.06ms or 21.79ms, peak memory 6.96GB.

File: benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_training.csv [hunk 1]
```diff
@@ -114,7 +114,7 @@ hf_Reformer,pass,23
 
 
 
-hf_Roberta_base,fail_accuracy,6
```
Comment (shunting314) [inline benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_training.csv:117]: I can not repro the failure locally. Also from the CI logs, it does not show which tensor has violates the tolerance. It may not be real accuracy problem
